text/0066-separate-context-propagation.md

```diff
@@ -0,0 +1,557 @@
+* [Internal details](#Internal-details)
+* [FAQ](#faq)
+
+Refactor OpenTelemetry into a set of separate cross-cutting concerns which 
+operate on a shared context propagation mechanism.
+
+# Motivation
+
+This RFC addresses the following topics:
+
+**Separation of concerns**  
+* Remove the Tracer dependency from context propagation mechanisms.
+* Handle user data (Correlations) and observability data (SpanContext, etc) 
+  separately.  
+
+**Extensibility**
+* Allow developers to create new applications for context propagation. For 
+  example: A/B testing, encrypted or authenticated data, and new, experimental 
+  forms of observability.
+
+
+# OpenTelemetry layered architecture
+
+The design of OpenTelemetry is based on the principles of [aspect-oriented 
+programming](https://en.wikipedia.org/wiki/Aspect-oriented_programming), 
+adopted to the needs of distributed systems. 
+
+Some concerns "cut across" multiple abstractions in a program. Logging 
+exemplifies aspect orientation because a logging strategy necessarily affects 
+every logged part of the system. Logging thereby "cross-cuts" across all logged 
+classes and methods. Distributed tracing takes this strategy to the next level, 
+and cross-cuts across all classes and methods in all services in the entire 
+transaction. This requires a distributed form of the same aspect-oriented 
+programming principles in order to be implemented cleanly.
+
+OpenTelemetry approaches this by separating it's design into two layers. The top 
+layer contains a set of independent **cross-cutting concerns**, which intertwine 
+with a program's application logic and cannot be cleanly encapsulated. All 
+concerns share an underlying distributed **context propagation** layer, for 
+storing state and accessing data across the lifespan of a distributed 
+transaction.
+
+
+# Cross-Cutting Concerns
+
+## Observability API
+Distributed tracing is one example of a cross-cutting concern. Tracing code is 
+interleaved with regular code, and ties together independent code modules which 
+would otherwise remain encapsulated. Tracing is also distributed, and requires 
+non-local, transaction-level context propagation in order to execute correctly.
+
+The various observability APIs are not described here directly. However, in this new 
+design, all observability APIs would be modified to make use of the generalized 
+context propagation mechanism described below, rather than the tracing-specific 
+propagation system it uses today.
+
+Note that OpenTelemetry APIs calls should *always* be given access to the entire 
+context object, and never just a subset of the context, such as the value in a 
+single key. This allows the SDK to make improvements and leverage additional 
+data that may be available, without changes to all of the call sites.
+
+## Correlations API
+
+In addition to trace propagation, OpenTelemetry provides a simple mechanism for 
+propagating metric labels, called the Correlations API. Correlations are 
```

Comment:
not only metric labels. Those may also label spans and logs

Reply:
Gotcha

Reply:
Please fix this :)